library/blocks/in-world-recipes/src/main/java/org/quiltmc/qsl/inworldrecipes/api/InWorldRecipeRegistries.java

```diff
@@ -0,0 +1,42 @@
+package org.quiltmc.qsl.inworldrecipes.api;
+
+import com.mojang.serialization.Lifecycle;
+import net.minecraft.block.Block;
+import net.minecraft.item.ItemUsageContext;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryKey;
+import net.minecraft.util.registry.SimpleRegistry;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Optional;
+
+/**
+ * <b>NOTE:</b> You need to handle action sound playback yourself.<br>
+ * <b>NOTE2:</b> You <em>can</em> override vanilla recipes with this.
+ */
+public final class InWorldRecipeRegistries {
+	private InWorldRecipeRegistries() { throw new AssertionError(); }
+
+	public static final Registry<InWorldRecipe> SWORD = create("sword");
+	public static final Registry<InWorldRecipe> PICKAXE = create("pickaxe");
+	public static final Registry<InWorldRecipe> AXE = create("axe");
+	public static final Registry<InWorldRecipe> SHOVEL = create("shovel");
+	public static final Registry<InWorldRecipe> HOE = create("hoe");
+
+	public static @NotNull Optional<InWorldRecipe> findMatchingRecipe(@NotNull Registry<InWorldRecipe> registry,
+			@NotNull ItemUsageContext context, @NotNull Block targetBlock) {
+		return registry.stream()
+				.filter(recipe -> recipe.targetBlock() == targetBlock && recipe.predicate().test(context))
+				.findFirst();
+	}
```

Comment:
Would this method ever be called in hot code? As it stands right now this is a pretty allocation-heavy and slow way of doing things, but that might be fine.

Reply:
I could probably transform this into a regular `for (Entry : registry.getEntries)` (especially since I actually need to change it to return a list).

Reply:
This definitely needs to be a `for` because the stream API is a disappointment so it will actually go through the entire list and never return early despite you needing just the first match.

Reply:
> I actually need to change it to return a list

Couldn't you add a `doRecipes` method which would try all matching (by block) recipes until one succeeds?

Reply:
> Couldn't you add a `doRecipes` method which would try all matching (by block) recipes until one succeeds?

What a coincidence, that's what I just did 🙃